src/session.c

```diff
@@ -492,18 +504,15 @@ static CK_RV session_pool_init(P11PROV_CTX *ctx, CK_TOKEN_INFO *token,
     if (token->ulMaxSessionCount != CK_EFFECTIVELY_INFINITE
         && token->ulMaxSessionCount != CK_UNAVAILABLE_INFORMATION) {
         pool->max_sessions = token->ulMaxSessionCount;
-        /* keep a max of 10% of the sessions */
-        pool->limit_sessions = pool->max_sessions / 10;
-        if (pool->limit_sessions == 0) {
-            pool->limit_sessions = 3;
-        }
-    } else {
-        /* arbitrary max concurrent open sessions */
+    }
+    /* arbitrary max concurrent open sessions */
+    if (pool->max_sessions <= 0 || pool->max_sessions > 1024) {
         pool->max_sessions = 1024;
-        pool->limit_sessions = 8;
     }
-    if (pool->limit_sessions > pool->max_sessions) {
-        pool->limit_sessions = pool->max_sessions;
+    /* arbitrary max concurrent cached sessions */
+    pool->max_cached_sessions = 8;
+    if (pool->max_cached_sessions > pool->max_sessions) {
+        pool->max_cached_sessions = pool->max_sessions << 8;
```

Comment:
this makes max cached sessions 4 for max_sessions 1024, which sounds very few. Is this really the intention or should we keep something more similarly as prior this change?

Reply:
Not intentional, I was thinking that I would like to have 8 as the result in the default case, then probably got distracted and wrote 8 in the shifting instead of a more appropriate 7 ... or maybe even 6.

I think this code is just too naive tbh, I would love to have a much system where we keep some data on average and peak number of concurrent open operations and then make a decision that make sense given those values. Even 8 may just be a waste if it is just a peak, and the program normally have 2 concurrent operations at most.

I will change this value to 7 now so that defaults and behavior follow the same ratios.

Reply:
Think of it, 7 makes no sense either, given here we are in the case where max_session is < 8, so I am just going to use the old logic. I think this whole heuristic will have to be reviewed in another patch